File: leetcode/problems/language-specific/go/399.evaluate-division.go
/*
 * @lc app=leetcode id=399 lang=golang
 *
 * [399] Evaluate Division
 */

// @lc code=start
var exists = struct{}{}

type Node struct {
	Key string
	Val float64
}

func calcEquation(equations [][]string, values []float64, queries [][]string) []float64 {
	graph := buildGraph(equations, values)
	res := make([]float64, len(queries))
	for i := 0; i < len(queries); i++ {
		src, dest := queries[i][0], queries[i][1]
		res[i] = calcEquationHelperDFS(src, dest, map[string]struct{}{}, graph)
	}

	return res
}

func buildGraph(equations [][]string, values []float64) map[string][]Node {
	graph := map[string][]Node{}
	for i := 0; i < len(equations); i++ {
		src, dest := equations[i][0], equations[i][1]
		graph[src] = append(graph[src], Node{dest, values[i]})
		graph[dest] = append(graph[dest], Node{src, 1 / values[i]})
	}

	return graph
}

func calcEquationHelperDFS(src, dest string, visited map[string]struct{}, graph map[string][]Node) float64 {
	_, srcOk := graph[src]
	_, destOk := graph[dest]
	if !srcOk && !destOk {
		return -1.0
	}
	if src == dest {
		return 1.0
	}

	visited[src] = exists
	for _, v := range graph[src] {
		if _, visitedOk := visited[v.Key]; !visitedOk {
			res := calcEquationHelperDFS(v.Key, dest, visited, graph)
			if res != -1.0 {
				return res * v.Val
			}
		}
	}

	return -1.0
}

// @lc code=end

File: leetcode/problems/sessions/2023-05-01_2023-09-16/399.evaluate-division.py
#
# @lc app=leetcode id=399 lang=python3
#
# [399] Evaluate Division
#

# @lc code=start

# * Iterative BFS Solution | O(n*(v+e)) Time | O(v+e) Space
# * v -> The number of vertices in the graph | e -> The number of edges in the graph |
# * n -> The length of queries array


class Solution:
    def calcEquation(
        self, equations: List[List[str]], values: List[float], queries: List[List[str]]
    ) -> List[float]:
        graph = defaultdict(list)

        def build_graph():
            for i, equation in enumerate(equations):
                src, dest = equation
                graph[src].append((dest, values[i]))
                graph[dest].append((src, 1 / values[i]))

        def bfs(src, dest):
            if src not in graph or dest not in graph:
                return -1

            queue = deque([(src, 1)])
            visited_set = set([src])
            while queue:
                cur_node, cur_val = queue.popleft()
                if cur_node == dest:
                    return cur_val

                for neighbor, neighbor_val in graph[cur_node]:
                    if neighbor not in visited_set:
                        queue.append((neighbor, cur_val * neighbor_val))
                        visited_set.add(neighbor)
            return -1

        build_graph()
        return [bfs(src, dest) for src, dest in queries]


# @lc code=end
